File: app/src/main/java/com/e/baking/IngredientDetailsRecyclerViewAdapter.java
package com.e.baking;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.e.baking.model.Ingredient;
import com.e.baking.model.Recipe;

import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;

public class IngredientDetailsRecyclerViewAdapter extends RecyclerView.Adapter<IngredientDetailsRecyclerViewAdapter.ViewHolder> {


    private int mNumberOfItems;
    private List<Ingredient> ingredients;
    private LayoutInflater mLayoutInflater;
    private RecipeRecyclerViewAdapter.ItemClickListener mItemClickListener;
    private Context context;

    public IngredientDetailsRecyclerViewAdapter(List<Ingredient> ingredients, Context context, int numberOfItems, BtnClickListener btnClickListener){
        this.ingredients=ingredients;
        this.mNumberOfItems=numberOfItems;
        this.mLayoutInflater=LayoutInflater.from(context);
        this.context=context;
    }


    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view =  LayoutInflater.from(parent.getContext()).inflate(R.layout.linearlayout_ingredients,parent,false);
        return new IngredientDetailsRecyclerViewAdapter.ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder viewHolder, int position) {
        Ingredient ingredient = this.ingredients.get(position);

        viewHolder.mNameTextView.setText(ingredient.getIngredient());
        viewHolder.mQuantityTextView.setText(ingredient.getQuantity());
    }

    @Override
    public int getItemCount() {
        return this.mNumberOfItems;
    }

    public class ViewHolder extends RecyclerView.ViewHolder implements RecyclerView.OnClickListener{
        @BindView(R.id.ingredient_name)
        TextView mNameTextView;
        @BindView(R.id.ingredient_quantity)
        TextView mQuantityTextView;


        public ViewHolder(@NonNull View itemView) {
            super(itemView);
            ButterKnife.bind(this, itemView);
            itemView.setOnClickListener(this);
        }

        @Override
        public void onClick(View view) {
            if(mItemClickListener != null) mItemClickListener.onItemClick(view,getAdapterPosition());
        }

    }

    public void setClickListener(RecipeRecyclerViewAdapter.ItemClickListener itemClickListener){
        this.mItemClickListener=itemClickListener;
    }

    public interface ItemClickListener {
        void onItemClick(View view, int position);
    }


    public interface BtnClickListener {
        public abstract void onBtnClick(int position);
    }
}

File: app/src/main/java/com/e/baking/IngredientDetailsRecyclerViewAdapter.kt
package com.e.baking

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import butterknife.BindView
import butterknife.ButterKnife
import com.e.baking.model.Ingredient

class IngredientDetailsRecyclerViewAdapter(
    private val ingredients: List<Ingredient>,
    context: Context,
    private val mNumberOfItems: Int,
    btnClickListener: BtnClickListener
) : RecyclerView.Adapter<IngredientDetailsRecyclerViewAdapter.ViewHolder>() {

    private var mItemClickListener: RecipeRecyclerViewAdapter.ItemClickListener? = null

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val view = LayoutInflater.from(parent.context)
            .inflate(R.layout.linearlayout_ingredients, parent, false)
        return ViewHolder(view)
    }

    override fun onBindViewHolder(viewHolder: ViewHolder, position: Int) {
        val ingredient = ingredients[position]
        viewHolder.mNameTextView?.text = ingredient.ingredient
        viewHolder.mQuantityTextView?.text = ingredient.quantity
    }

    override fun getItemCount(): Int = mNumberOfItems

    fun setClickListener(itemClickListener: RecipeRecyclerViewAdapter.ItemClickListener) {
        mItemClickListener = itemClickListener
    }

    inner class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView), View.OnClickListener {
        @JvmField
        @BindView(R.id.ingredient_name)
        var mNameTextView: TextView? = null

        @JvmField
        @BindView(R.id.ingredient_quantity)
        var mQuantityTextView: TextView? = null

        init {
            ButterKnife.bind(this, itemView)
            itemView.setOnClickListener(this)
        }

        override fun onClick(view: View) {
            mItemClickListener?.onItemClick(view, adapterPosition)
        }
    }

    interface ItemClickListener {
        fun onItemClick(view: View, position: Int)
    }

    interface BtnClickListener {
        fun onBtnClick(position: Int)
    }
}
